lib/carconntwistprim.cc

```diff
@@ -0,0 +1,270 @@
+#include <carconntwistprim.h>
+#include <iostream>
+#include "utils.h"
+#include <algorithm>
+
+
+namespace dsl {
+
+using namespace Eigen;
+using namespace std;
+
+using Eigen::Vector2d;
+using Eigen::Vector3d;
+using Eigen::Vector3d;
+using Eigen::Matrix3d;
+using Eigen::Affine3d;
+using std::vector;
+
+typedef Eigen::Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> MatrixXbool;
+
+CarConnTwistPrim::CarConnTwistPrim(const CarGrid& grid, CarPrimitiveCfg& cfg)
+:grid(grid){
+  SetPrimitives(1, cfg);
+}
+
+Vector2d xy2wt2( double xf,double yf, double u ){
+  double w, t;
+  if(abs(yf)<1e-12){
+    w=0;
+    t=xf/u;
+  }else{
+    w = 2*yf/(u*(xf*xf+yf*yf));
+    t = atan2(w*xf/u, 1-w*yf/u)/w;
+  }
+  return Vector2d(w,t);
+}
+
+CarConnTwistPrim::CarConnTwistPrim(const CarGrid& grid,
+                                 const vector<Eigen::Vector3d> &vs,
+                                 double dt) : grid(grid), vs(vs), dt(dt)
+{
+}
+
+CarConnTwistPrim::CarConnTwistPrim(const CarGrid& grid,
+                                 double dt,
+                                 double vx,
+                                 double kmax,
+                                 int kseg,
+                                 bool onlyfwd)
+: grid(grid) {
+  SetPrimitives(dt, vx, kmax, kseg, onlyfwd);
+}
+
+bool CarConnTwistPrim::SetPrimitives(double dt, CarPrimitiveCfg& cfg){
+
+  cfg.nl = cfg.nl<2 ? 2: cfg.nl;  //! Make sure the number of different traj length is atleast 2
+  cfg.na = cfg.na<3 ? 3: cfg.na;  //! Make sure the number of differentsteering angle is atleast 3
+
+  double u = 1.0; //The speed can be anything. Using for clarity
+
+  double tmin = cfg.lmin/u;
+  double tmax = cfg.lmax/u;
+  double wmax = u*cfg.tphioverlmax;
+  double rmax = cfg.lmax/cfg.amax;
+  double ymax = rmax - rmax*cos(cfg.amax);
+  double xmax = cfg.lmax;
+
+  this->dt = dt;
+
+  double del_t = exp(log(tmax/tmin)/(cfg.nl-1));
+
+  double del_w = wmax/(cfg.na-1);
+
+  double sx = grid.cs(1);
+  double sy = grid.cs(2);
+
+  //create a grid which can accomodate any primitive starting from any orientation
+  uint grid_nhc = ceil(max(ymax/sy, xmax/sx)); // number of half cells
+  uint grid_nc  = 2*grid_nhc + 1;
+  Matrix<bool, Dynamic, Dynamic> grid_seen(grid_nc, grid_nc);
+
+  Affine3d igorg_to_mgcorg = Scaling(Vector3d(1/sx, 1/sy, 1)) *Translation3d(Vector3d(grid_nhc*sx, grid_nhc*sx, 0));
+
+  //Allocate space for the
+  vss.resize(grid.gs(0));
+  for(size_t i=0;i<vss.size();i++)
+    vss[i].reserve(cfg.na*cfg.nl);
+
+  if(cfg.tocenter){
+    for(int idx_a=0; idx_a < grid.gs(0);idx_a++){
```

Comment:
need documentation on what is happening here, otherwise it's difficult to follow